Update to BoringSSL `fips-20220613` tag

reedloden opened the pull request:
Matches what upstream Go is using for BoringCrypto build: https://github.com/golang/go/commit/7383b2a4db5dc93c9b875b42d5add73d27cc4b9f.

As this is not yet upstreamed, a separate fork is being used -- https://github.com/gravitational/boring.

Additionally, in our build pipeline:
- Updates clang to version 14.0.6, in line with what Go is using
- Sets `CMAKE=cmake3`, as `boring-rs` (by way of `cmake-rs`) calls `cmake` (which isn't installed).
- Installs `ninja`, which is the FIPS-approved build tool for boring crypto. (Incidentally the `make`-built version of boring crypto also appears to be broken).

reedloden committed:
Update Clang from 12 to 14 to match BoringSSL `fips-20220613` tag

Matches what upstream Go is using for BoringCrypto build:
https://github.com/golang/go/commit/7383b2a4db5dc93c9b875b42d5add73d27cc4b9f

Also set `CMAKE=cmake3`, as `boring-rs` (by way of `cmake-rs`) calls `cmake`.

Additionally, make `build-centos7-assets` a dependency.

fheinecke commented on build.assets/Dockerfile-centos7-assets:
```diff
@@ -36,22 +36,22 @@ RUN yum install --nogpgcheck -y \
 # This version of Clang is explicitly required for FIPS compliance when building BoringSSL.
 # For more information please refer to the section 12. Guidance and Secure Operation of:
-# https://csrc.nist.gov/CSRC/media/projects/cryptographic-module-validation-program/documents/security-policies/140sp4407.pdf
+# TODO(reed) -- fill in URL
```
Is this TODO for later or does it need to b a part of this PR?

reedloden replied:
There is no updated security policy available yet, so it's still TODO for some future point.

fheinecke commented on build.assets/Dockerfile-centos7-assets:
```diff
@@ -36,22 +36,22 @@ RUN yum install --nogpgcheck -y \
 ARG DEVTOOLSET
 
-# Compile Clang 12.0.0 from source. It is needed to create BoringSSL and BPF files.
+# Compile Clang 14.0.6 from source. It is needed to create BoringSSL and BPF files.
```
Can you cut/publish a dev tag prior to merge? Jumping 2 major versions of clang could potentially cause build issues and/or cause unexpected bugs.

IMO this is major enough that we should go through the whole test plan with the version bumped, but that's probably not feasible at this stage.

reedloden replied:
Dev tag won't actually test any of this, as the assets build is separate from the build pipeline and buildbox usage.

But that said, aren't we only using clang for BoringSSL and libbpf? Is there anything else that uses clang? I don't see why we would need to redo the entire test plan. At most, this would affect FIPS builds and BPF features.

fheinecke replied:
Ah alright. I was under the impression that we used Clang during much more of the build process. Do you know if we're actually shipping these built libbpf and boringssl libs to customers? If so then it'd still be nice to have some tests ran against the final build, but it's way lower risk than I originally thought.

jakule committed:
Add ninja-build setup to Dockerfile-centos7

The Dockerfile-centos7 has been updated to include the setup and installation of ninja-build. The version of ninja-build included in CentOS 7 AMR64 was too old for building boringssl, hence it needs to be built from source. Both additional dependencies and the source build steps have been added as part of this update.

jakule committed:
Add CMake to the ninja build

jakule committed:
Update Dockerfile to fix ARM64 build issue

The Dockerfile for Centos7 was updated to better support ARM64 architectures. Instead of fetching a precompiled binary of CMake, we now clone CMake v3.28.1 and compile it specifically for the target. This change includes the addition of the `-mno-outline-atomics` flag, which resolves ARM64-specific issues with compilation.

jakule committed:
Update Dockerfile to conditionally set CFLAGS for arm64 architecture

The Dockerfile for CentOS7 now includes a condition that sets the CFLAGS environment for arm64 architecture. This change was made during the process of enabling CMake to bootstrap itself and ensures that correct flags are set for specific architectures during the build process.

Commit:
Removes build-centos7-assets as a dependency for the build-centos7 target.
Adds a comment to the Makefile to explain why.

ibeckermayer commented on build.assets/Makefile:
```diff
@@ -160,7 +160,7 @@ buildbox-fips: buildbox-centos7-fips
 #
 .PHONY:buildbox-centos7
-buildbox-centos7:
+buildbox-centos7: build-centos7-assets
```
This was a suggestion of mine, but I understand your point. I removed it and added an explanatory comment: 5689727e1c13e8c5f476bda717ca0c50d2930ee5

Commit:
Updated Clang build configuration in Dockerfile

Updates clang to be built with the inclusion of the header files
necessary for building boring. Also changes the build to use Ninja.

Commit:
moves the ninja copy closer to the clang copy

reedloden committed:
Swap to fork of `boring-rs` for build fixes (#36720)

BoringSSL does not build properly on CentOS 7 without some build fixes. As these fixes are not yet upstreamed, a separate fork is being used -- https://github.com/gravitational/boring.

Co-authored-by: Isaiah Becker-Mayer <isaiah@goteleport.com>

Commit:
Merge branch 'master' into reed/boringssl-fips-20220613

Commit:
Ninja build snuck back into Dockerfile-centos7, but we don't need it since it's in Dockerfile-centos7-assets now. Removing it.

jakule committed:
Prevent docker buildkit from caching

Use docker build to build fips image to prevent docker from fetching the base centos image from the registry.

jakule committed:
Add 'nolint' comment in boringtest.go

A 'nolint' comment was added in the boringtest go file to ignore linting this file

jakule committed:
Rename test Go file
